README

		Memory Channels in Tcl
		(Version 2.3 / DEC-10-2010)

	Andreas Kupries	(andreas_kupries@users.sourceforge.net)

This directory contains a freely distributable extension to Tcl
implementing memory channels, i.e. channels storing the data
placed into them in memory, not on disk.


BACKGROUND

Tcl stands for the Tool Command Language, a freely distributable,
embeddable scripting language package.  Tk is an freely distributable X
windows interface toolkit and widget library implemented as a collection of
new commands added to Tcl.

Memchan 2.3 is developed for Tcl 8.0 or higher.  It is a Tcl
extension, Tk is not required.

This distribution contains the source code for Memchan + (very simple)
documentation describing the new command.
Memchan should be available by anonymous ftp from ftp.neosoft.com soon.


COMPILING/USING Memchan

To compile Memchan follow the instructions in the file README.install

To report bugs, bug fixes, descriptions of interesting Memchan
applications, and suggested improvements:

+) Send email to andreas_kupries@users.sourceforge.net

or

+) Post an article to the newsgroup	comp.lang.tcl


REPORTING PROBLEMS

This extension has been tested extensively on a number of platforms and with
various versions of Tcl. If you find a problem and can reproduce it simply
then please submit a bug report on the memchan sourceforge project at
http://sourceforge.net/projects/memchan/


ChangeLog
2011-02-14  Andreas Kupries  <andreask@activestate.com>

	* generic/fifo.c (ChannelReady): [Bug 3175754]: Fixed inverted
	condition killing readable events exactly when they were needed.

2010-12-10  Andreas Kupries  <andreas_kupries@users.sourceforge.net>

	* *** Version 2.3 tagged and released ***

	* generic/memchanInt.h: Various fixes to unbreak the int2ptr
	* configure.in: introduction.
	* configure:

2010-12-08  Andreas Kupries  <andreask@activestate.com>

	* Makefile.in: Dropped bogus 'doc' dependency from 'dist' target.
	
	* generic/memchanInt.h: Swiped the INT2PTR/PTR2INT macros from
	* generic/buf.c: Tcl and updated the code handling AssocData to
	* configure.in: use them, kill a few warnings on 64bit systems.
	* configure: Plus related auto-goo.

	* generic/memchan.c: Second set of patches by Stuart. Moved
	* generic/memchanInt.h: string.h inclusion to internal header,
	* Makefile.in: and moved the .SUFFIXES declaration to a place
	  where the OpenBSD make can use it.

	* doc/fifo.man: Applied patch supplied by Stuart Cassoff fixing
	* doc/fifo2.man: various typos in the manpages.
	* doc/memchan.man:
	* doc/null.man:
	* doc/zero.man:

2010-12-06  Andreas Kupries  <andreas_kupries@users.sourceforge.net>

	* Dropped the old build files under unix/ and win/ in favor of the
	  toplevel TEA files.

	* configure.in:  Updated to TEA 3.9.
	* tclconfig/README.txt:
	* tclconfig/tcl.m4:
	* configure:

	* ANNOUNCE: Version bumped to 2.3.
	* ANNOUNCE.MAIL:
	* ChangeLog:
	* ChangeLog.short:
	* DESCRIPTION:
	* LSM:
	* README:
	* teapot.txt:

	* Makefile.in: Updated the dist target.

2009-03-16  Andreas Kupries  <andreask@activestate.com>

	* generic/bugExt.c: Fixed [SF Bug 2687845] by Stuart Cassoff,
	* generic/bugFix.c: added inclusion of memchanInt.h to these
	* generic/bugRange.c: files to declare memcpy().

2008-10-03  Andreas Kupries  <andreask@activestate.com>

	* generic/memchanInt.h: Fixed [SF Bug 2078168] by Larry Virden
	  <lvirden@users.sourceforge.net>. Added inclusion of header
	  <string.h> to get declarations for strcmp, memcpy, and memset.

2007-11-13  Andreas Kupries  <andreask@activestate.com>

	* teapot.txt: New file, basic meta data of the package for TEApot
	  compatible repositories.

2006-05-08  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* Makefile.in:      Fixed make test target.

2006-01-24  Andreas Kupries <andreask@activestate.com>

	* tclconfig/tcl.m4: Updated to TEA 3.5
	* configure.in:     Ditto.
	* configure:        Regenerated.

2005-06-08  Andreas Kupries <andreask@activestate.com>

	* Makefile.in: Updated the toplevel build system to TEA 3.2.
	* configure.in:
	* generic/init.c:
	* generic/memchanStubLib.c:
	* tclconfig/tcl.m4:

2004-11-10  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* Makefile.in: Fix the unix build system for the stubs
	stuff. Verified that the generated stubs lib can be used in an
	extension (make chantest target).

2004-11-10  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* doc/memchanapi.man: NEW documentation for the stubs API functions.

	* generic/memchan.c:  Fix and test for bug #1060620 which shows a
	* tests/memchan.test: problem using fileevent on memchan channels

	* generic/fifo2.c: Added special case for attempts to build on
	threaded tcl 8.2. This version doesn't support Tcl_MutexFinalize
	and we would leak mutexes from fifo2. Provided error stub
	functions and included a #error so that anyone attempting to build
	under these conditions will know what the problem is.

	* generic/null.c:        Fix for compilation with tcl 8.2 and above.
	* generic/random.c:
	* generic/zero.c:
	* generic/memchanInt.h:

2004-11-09  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* generic/memchan.h:          Added a stubs library for the Memchan
	* generic/memchan.decls:      extension. This exposes an API that
	* generic/memchanDecls.h:     permits C extensions to create Memchan
	* generic/memchanStubInit.c:  channels directly.
	* generic/memchanStubLib.c:
	
	* generic/chantest.c:     Sample extension for testing the stubs
	                          linking C API.
	
	* doc/manpage.css:        Added the tcllib style-sheet.

	* generic/memchan.c:      Modified in support of the stubs API. 
	* generic/fifo.c:         Each channel creation command now uses
	* generic/fifo2.c:        the C API internally. Channel creation code
	* generic/null.c:         moved to a function exposed on the Memchan
	* generic/zero.c:         stubs interface.
	* generic/random.c:
	* generic/buf.decls:
	* generic/bufStubInit.c:

	* win/makefile.vc:        Windows build files updated
	* win/rules.vc:
	* configure.in:           TEA files updated with new version.
	* configure:

	==== Incremented version to 2.2.1 ====
	
2004-08-04  Pat Thoyts  <pat@zsplat.freeserve.co.uk>

	* generic/fifo2.c: Applied fix for bug #996078 which sets the channel
	                   mutex to NULL in threaded tcl (Jeff Gilbertson)

2004-06-04  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* configure:        Updated and regenerated the TEA configure script
	* tclconfig/tcl.m4: This now works with Msys on Windows (gcc & cl)

2004-06-04  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* generic/zero.c:   Make the interval between fileevent
	* generic/null.c:   notifications configurable (FR #470415).
	* generic/random.c:
	* doc/zero.man:     Add the -delay option to the manual pages.
	* doc/null.man:
	* doc/random.man:
	* tests/zero.test:  Additional tests for -delay ?ms? option.
	* tests/null.test:
	* tests/random.test:

2004-06-03  Pat Thoyts  <patthoyts@users.sourceforge.net>

	* configure.in:         Changed version to 2.2.0
	* configure:
	* DESCRIPTION:
	* win/makefile.vc:
	* win/mc.rc:
	* unix/Makefile.in:     Updated unix/* build files for the new
	* unix/configure.in:    random and zero channels.
	* unix/configure:

	* generic/zero.c:       Added a 'zero' channel (FR #470414)
	* tests/zero.test:
	* doc/zero.man:
	* generic/init.c:
	* generic/memchanInt.h:
	* Makefile.in:

generic/init.c
/* -*- c -*-
 * init.c --
 *
 *	Implements the C level procedures handling the initialization of
 *	this package
 *
 *
 * Copyright (c) 1996-1999 Andreas Kupries (a.kupries@westend.com)
 * Copyright (c) 2000-2005 Andreas Kupries (akupries@shaw.ca)
 * All rights reserved.
 *
 * Permission is hereby granted, without written agreement and without
 * license or royalty fees, to use, copy, modify, and distribute this
 * software and its documentation for any purpose, provided that the
 * above copyright notice and the following two paragraphs appear in
 * all copies of this software.
 *
 * IN NO EVENT SHALL I LIABLE TO ANY PARTY FOR DIRECT, INDIRECT, SPECIAL,
 * INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OF THIS
 * SOFTWARE AND ITS DOCUMENTATION, EVEN IF I HAVE BEEN ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * I SPECIFICALLY DISCLAIM ANY WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND
 * I HAVE NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
 * ENHANCEMENTS, OR MODIFICATIONS.
 *
 * CVS: $Id: init.c,v 1.10 2004/06/03 23:39:13 patthoyts Exp $
 */

/*#include <stdlib.h>*/
#include "memchanInt.h"
#include "buf.h"

extern BufStubs bufStubs;

char *
Buf_InitStubs _ANSI_ARGS_((Tcl_Interp *interp, CONST char *version, int exact));

#ifndef PACKAGE_VERSION
#define PACKAGE_VERSION MEMCHAN_VERSION
#endif
#ifndef PACKAGE_NAME
#define PACKAGE_NAME "Memchan"
#endif


/*
 *------------------------------------------------------*
 *
 *	Memchan_Init --
 *
 *	------------------------------------------------*
 *	Standard procedure required by 'load'. 
 *	Initializes this extension.
 *	------------------------------------------------*
 *
 *	Sideeffects:
 *		As of 'Tcl_CreateCommand'.
 *
 *	Result:
 *		A standard Tcl error code.
 *
 *------------------------------------------------------*
 */

int Memchan_Init (interp)
Tcl_Interp* interp;
{
#if GT81
  if (Tcl_InitStubs (interp, "8.1", 0) == NULL) {
    return TCL_ERROR;
  }
#endif

  Tcl_CreateObjCommand (interp, "memchan",
			&MemchanCmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

  Tcl_CreateObjCommand (interp, "fifo",
			&MemchanFifoCmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

  Tcl_CreateObjCommand (interp, "fifo2",
			&MemchanFifo2Cmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

  Tcl_CreateObjCommand (interp, "null",
			&MemchanNullCmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

  Tcl_CreateObjCommand (interp, "random",
			&MemchanRandomCmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

  Tcl_CreateObjCommand (interp, "zero",
			&MemchanZeroCmd,
			(ClientData) NULL,
			(Tcl_CmdDeleteProc*) NULL);

#if GT81
    /* register extension and its interfaces as now available package
     */
    Tcl_PkgProvideEx (interp, PACKAGE_NAME, PACKAGE_VERSION, (ClientData) &bufStubs);

#ifndef __WIN32__
    Buf_InitStubs (interp, PACKAGE_VERSION, 0);
#endif
#else
  /* register memory channels as available package */
  Tcl_PkgProvide (interp, PACKAGE_NAME, PACKAGE_VERSION);
#endif

  Buf_Init (interp);
  return TCL_OK;
}

/*
 *------------------------------------------------------*
 *
 *	Memchan_SafeInit --
 *
 *	------------------------------------------------*
 *	Standard procedure required by 'load'. 
 *	Initializes this extension for a safe interpreter.
 *	------------------------------------------------*
 *
 *	Sideeffects:
 *		As of 'Memchan_Init'
 *
 *	Result:
 *		A standard Tcl error code.
 *
 *------------------------------------------------------*
 */

int Memchan_SafeInit (interp)
Tcl_Interp* interp;
{
  return Memchan_Init (interp);
}

